modules/core/04-channel/keeper/packet.go

```diff
@@ -133,11 +134,16 @@ func (k Keeper) RecvPacket(
 		return errorsmod.Wrap(types.ErrChannelNotFound, packet.GetDestChannel())
 	}
 
-	if channel.State != types.OPEN {
-		return errorsmod.Wrapf(
-			types.ErrInvalidChannelState,
-			"channel state is not OPEN (got %s)", channel.State.String(),
-		)
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.TRYUPGRADE, types.ACKUPGRADE}) {
+		return errorsmod.Wrapf(types.ErrInvalidChannelState, "channel state was not one of [%s, %s, %s] (got %s)", types.OPEN.String(), types.TRYUPGRADE.String(), types.ACKUPGRADE.String(), channel.State.String())
+	}
+
+	// in the case of the channel being in TRYUPGRADE or ACKUPGRADE we need to ensure that the channel is not in flushing,
+	// and that the counterparty last sequence send is less than or equal to the packet sequence.
+	if counterpartyLastSequenceSend, found := k.GetCounterpartyLastPacketSequence(ctx, packet.GetDestPort(), packet.GetDestChannel()); found {
+		if channel.FlushStatus != types.FLUSHING || packet.GetSequence() > counterpartyLastSequenceSend {
+			return errorsmod.Wrapf(types.ErrInvalidChannelState, "expected channel flush status to not be (%s) and for the packet sequence (%d) to be less than or equal to the counterparty last sequence send (%d)", types.FLUSHING.String(), packet.GetSequence(), counterpartyLastSequenceSend)
```

Comment:
I wonder could we improve this error and msg a bit. I'm not sure if this is much better but it would be good to indicate that we expect the channel to be flushing when the counterparty last seq send is set.

`"expected channel flush status to be FLUSHING when counterparty last sequence send (100) is set, failed to recv packet at sequence (99)"`

we could also change to `ErrInvalidFlushStatus` maybe? what do you think?

```suggestion
			return errorsmod.Wrapf(types.ErrInvalidFlushStatus, "expected channel flush status to be (%s) when counterparty last sequence send (%d) is set, failed to recv packet (%d)", types.FLUSHING, counterpartyLastSequenceSend, packet.GetSequence())
```

Reply:
won't this fail if the channel is just in a normal OPEN state and NOTINFLUSH? am i misreading something

Reply:
like, we should only run this condition with the if condition of TRY and ACKUPGRADE

Reply:
ah, it's only run if the counterpartyLastSeq send is found, which would only happen if we are in the upgrade process I guess.